sdk/core/Azure.Core/src/Pipeline/HttpPipeline.cs

```diff
@@ -70,7 +109,28 @@ public ValueTask SendAsync(HttpMessage message, CancellationToken cancellationTo
         {
             message.CancellationToken = cancellationToken;
             AddHttpMessageProperties(message);
-            return _pipeline.Span[0].ProcessAsync(message, _pipeline.Slice(1));
+
+            if (message.Policies == null || message.Policies.Count == 0)
+            {
+                return _pipeline.Span[0].ProcessAsync(message, _pipeline.Slice(1));
+            }
+
+            return SendAsync(message);
+        }
+
+        private async ValueTask SendAsync(HttpMessage message)
```

Comment:
nit: Why do we have a new private method here, but incooporate it directly into `public void Send(HttpMessage message, CancellationToken cancellationToken)` as an else block?

Reply:
This one is **async** -- it prevents a state machine from being added on the hot path.

Reply:
I'm not sure I follow why the additional method call is better than adding this code directly in place of the `return SendAsync(message)` call in the SendAsync method above?

Reply:
Have a look at the different in the compiler output for `DoFastAsync` and `DoAsync` in [this example](https://sharplab.io/#v2:D4AQTAjAsAUCAMACEECsBuWsQGZlkQGFEBvWRCxcy3ZAFkQFkBDASwDsAKASmorJiUhiACIB7ACoALDgHMAggGcAnuwDGnAC4AnAK4BTbpkGUAvnyomKtAGrMANgZAA2UZJnsFK9ZwBGYsXtEKTFNAAVmTSluUgshOMpWADNEThDwyOiE/mzhcQAxZkVNJVUNI1zzK2FckAB2N1KfCuqq+OrbBydXAqKS73KLAWEaBvZ9AHdEO0d9Fx5jITbKC1oQAA5prrmesSbB6uGR5ABOZGcAOgAlXS4eAF4APhJTFqXYUyA).

Reply:
Chatted offline to fully understand. The reason for this pattern is 2 fold (documenting here so this can be added as a comment later).
- The per-request policy path must be async because we must await the call to ProcessAsync so that the ArrayPool rental can be returned in the finally block.
- Because this path must be async, the additional method is needed to avoid creation of the async state machine in the hot path.